core/commonMain/src/implementations/immutableList/PersistentVectorBuilder.kt

```diff
@@ -134,6 +156,110 @@ class PersistentVectorBuilder<E>(private var vector: PersistentList<E>,
+            // there is enough space in tail, add all elements to it
+            tail = copyToBuffer(makeMutable(tail), tailSize, elementsIterator)
+            size += elements.size
+        } else {
+            val buffersSize = (elements.size + tailSize - 1) / MAX_BUFFER_SIZE
+            val buffers = arrayOfNulls<Array<Any?>?>(buffersSize)
+
+            // fill remained space of tail
+            buffers[0] = copyToBuffer(makeMutable(tail), tailSize, elementsIterator)
+            // fill other buffers
+            for (index in 1 until buffersSize) {
+                buffers[index] = copyToBuffer(mutableBuffer(), 0, elementsIterator)
+            }
+
+            // add buffers to the root, rootShift is updated appropriately there
+            @Suppress("UNCHECKED_CAST")
+            root = addBuffers(root, rootSize(), buffers as Array<Array<Any?>>)
+            // create new tail and copy remained elements there
+            tail = copyToBuffer(mutableBuffer(), 0, elementsIterator)
+            size += elements.size
+        }
+
+        return true
+    }
+
+    private fun copyToBuffer(buffer: Array<Any?>, bufferIndex: Int, sourceIterator: Iterator<Any?>): Array<Any?> {
+        var index = bufferIndex
+        while (index < MAX_BUFFER_SIZE && sourceIterator.hasNext()) {
+            buffer[index++] = sourceIterator.next()
+        }
+        return buffer
+    }
+
+    /**
+     * Adds all buffers from [buffers] as leaf nodes to the [root].
+     * If the [root] has less available leaves for the buffers, height of the trie is increased.
+     *
+     * Returns root of the resulting trie.
+     */
+    private fun addBuffers(root: Array<Any?>?, rootSize: Int, buffers: Array<Array<Any?>>): Array<Any?> {
+        val buffersIterator = buffers.iterator()
+
+        var mutableRoot = when {
+            rootSize shr LOG_MAX_BUFFER_SIZE < 1 shl rootShift ->
+                // if the root trie is not filled entirely, fill it
+                placeBuffers(root, rootSize, rootShift, buffersIterator)
+            else ->
+                // root is filled entirely, make it mutable
+                makeMutable(root)
+        }
+
+        // here root is filled entirely or/and all buffers are already placed
+
+        while (buffersIterator.hasNext()) {
+            // some buffers left, so root is filled entirely. promote root to the next level
+            rootShift += LOG_MAX_BUFFER_SIZE
+            mutableRoot = mutableBufferWith(mutableRoot)
+
+            placeBuffers(mutableRoot, 1 shl rootShift, rootShift, buffersIterator)
+        }
+
+        return mutableRoot
+    }
+
+    /**
+     * Places buffers from the [buffersIterator] as the leaf nodes.
+     * In the result of this call [root] node gets entirely filled, or all buffers get placed.
+     *
+     * Returns the resulting root.
+     */
+    private fun placeBuffers(root: Array<Any?>?, rootSize: Int, shift: Int, buffersIterator: Iterator<Array<Any?>>): Array<Any?> {
```

Comment:
Perhaps not the best name to describe what this functions does. Do I get it right, that is essentially the same as `pushTail` but for multiple buffers?

Reply:
Yes, it pushes several buffers to `root`.
Renamed to `pushBuffers`.